File: dining-philosophers/dp.py
# This is an implementation of Tanenbaum's solution to the Dining Philosophers problem from the Little Book of Semaphores.
import threading
import time
import random

NUM_TRIALS = 100
NUM_PHILOSOPHERS = 5
MAX_HELPINGS = 100 # Maximum number of times each philosopher eats.

waitingTime = [0]*NUM_PHILOSOPHERS

def rightNeighbour(i):
    return (i-1) % NUM_PHILOSOPHERS

def leftNeighbour(i):
    return (i+1) % NUM_PHILOSOPHERS

def getForks(i, state, sem, mutex):
    mutex.acquire()
    state[i] = "hungry"
    #print("Philosopher", i, "is hungry.")
    test(i, state, sem)
    mutex.release()
    sem[i].acquire()

def putForks(i, state, sem, mutex):
    mutex.acquire()
    state[i] = "thinking"
    #print("Philosopher", i, "is thinking.")
    test(leftNeighbour(i), state, sem)
    test(rightNeighbour(i), state, sem)
    mutex.release()

def test(i, state, sem):
    if state[i] == "hungry" and state[leftNeighbour(i)] != "eating" and state[rightNeighbour(i)] != "eating":
        state[i] = "eating"
        #print("Philosopher", i, "is eating.")
        sem[i].release()

def philosopher(i, state, sem, mutex):
    global waitingTime
    numHelpings = 0
    while numHelpings < MAX_HELPINGS:
        #time.sleep(0.001*random.randint(0,100)) # Simulate thinking.
        startWaiting = time.time()
        getForks(i, state, sem, mutex)
        doneWaiting = time.time()
        #time.sleep(0.001*random.randint(0,100)) # Simulate eating.
        putForks(i, state, sem, mutex)
        numHelpings += 1
        waitingTime[i] += doneWaiting-startWaiting

def main():
    durations = [0]*NUM_TRIALS
    averageWaitingTime = [0]*NUM_TRIALS
    for j in range(NUM_TRIALS):
        state = ["thinking"]*NUM_PHILOSOPHERS
        sem = [threading.Semaphore(0) for i in range(NUM_PHILOSOPHERS)]
        mutex = threading.Lock()
        philosophers = []
        start = time.time()
        for i in range(NUM_PHILOSOPHERS):
                philosopherThread = threading.Thread(target=philosopher, args=(i, state, sem, mutex))
                philosophers.append(philosopherThread)
                philosopherThread.start()

        for philosopherThread in philosophers:
                philosopherThread.join()

        end = time.time()
        global waitingTime
        durations[j] = end-start
        averageWaitingTime[j] = sum(waitingTime)/len(waitingTime)
        waitingTime = [0]*NUM_TRIALS

    print("Average waiting time:", sum(averageWaitingTime)/len(averageWaitingTime), "s")
    print("Average duration:", sum(durations)/len(durations), "s")    

if __name__ == "__main__":
    main()

File: dining-philosophers/dp.go
// This is an implementation using Go channels of the Dining Philosophers Solution #1 from the Little Book of Semaphores.
package main

import (
	"fmt"
	"sync"
	"time"
)

const numPhilosophers = 5
const maxHelpings = 100 // Maximum number of times each philosopher eats.
const numTrials = 100

var waitingTimes [numPhilosophers]time.Duration

func rightFork(i int) int {
	return i
}

func leftFork(i int) int {
	return (i + 1) % numPhilosophers
}

func getForks(i int, rightFork chan int, leftFork chan int) {
	rightFork <- i
	leftFork <- i
}

func putForks(i int, rightFork chan int, leftFork chan int) {
	<-rightFork
	<-leftFork
}

func philosopher(i int, table chan int, rightFork chan int, leftFork chan int, wg *sync.WaitGroup) {
	numHelpings := 0
	for numHelpings < maxHelpings {
		//time.Sleep(time.Duration(rand.Intn(100)) * time.Millisecond) // Simulates time to think.
		//fmt.Println("Philosopher", i, "is hungry")
		startWaiting := time.Now()
		table <- i
		getForks(i, rightFork, leftFork)
		endWaiting := time.Now()
		//fmt.Println("Philsopher", i, "is eating")
		//time.Sleep(time.Duration(rand.Intn(100)) * time.Millisecond) // Simulates time to eat.
		putForks(i, rightFork, leftFork)
		//fmt.Println("Philosopher", i, "is thinking")
		<-table
		numHelpings++
		waitingTimes[i] += endWaiting.Sub(startWaiting)
	}
	wg.Done()
}

func main() {
	var durations float64
	var averageWaitingTimes float64
	for j := 0; j < numTrials; j++ {
		var wg sync.WaitGroup
		table := make(chan int, numPhilosophers-1) // Makes sure only numPhilosophers-1 people are ever seated at the table at the same time (to avoid deadlock)
		var forks [numPhilosophers]chan int
		for i := range forks {
			forks[i] = make(chan int, 1)
		}
		start := time.Now()
		wg.Add(numPhilosophers)
		for i := 0; i < numPhilosophers; i++ {
			go philosopher(i, table, forks[rightFork(i)], forks[leftFork(i)], &wg)
		}
		wg.Wait()
		end := time.Now()
		durations += end.Sub(start).Seconds()
		var totalWaitingTime time.Duration
		for i := 0; i < numPhilosophers; i++ {
			totalWaitingTime += waitingTimes[i]
			waitingTimes[i] = 0
		}
		averageWaitingTimes += totalWaitingTime.Seconds() / numPhilosophers
	}
	fmt.Println("Average waiting time:", averageWaitingTimes/numTrials, "s")
	fmt.Println("Average duration:", durations/numTrials, "s")
}
